src/libraries/System.Linq/src/System/Linq/OrderedEnumerable.cs

```diff
@@ -155,6 +155,81 @@ internal override CachingComparer<TElement> GetComparer(CachingComparer<TElement
                 : new CachingComparerWithChild<TElement, TKey>(_keySelector, _comparer, _descending, childComparer);
             return _parent != null ? _parent.GetComparer(cmp) : cmp;
         }
+
+        internal override IEnumerator<TElement> GetEnumerator(int minIdx, int maxIdx)
+        {
+            // TODO: refine & benchmark different heuristics, enable custom comparer support?
+            return _parent is null && _comparer == Comparer<TKey>.Default && maxIdx < 1024
```

Comment:
I'll look in more depth later, but high level I expect this heuristic would need to change significantly for it to be valuable. I think it really needs to be about the ratio between the number of items taken and the number of items being sorted, eg only use this if a) we can get the actual number of items in the source enumerable cheaply and b) numberTaken / numberTotal < some threshold, like 50%.

Reply:
Just to avoid confusion, this is by no means the version I expect to be final. I made the draft PR with regards to https://github.com/dotnet/runtime/issues/96277#issuecomment-1896647794 and on https://github.com/dotnet/runtime/issues/96277#issuecomment-1898334685 how I can profile because the numbers are not what I expect...

What I have in mind is: if it is a very low number that we take or the ratio, if available, is favorable. But before testing that, I need to have confidence in the numbers I'm seeing...

Reply:
> how I can profile

Easiest way is writing a simple console app and using the VS profiler.

> it is a very low number

That will add non-trivial overhead (eg allocating the PriorityQueue) for an operation that will otherwise be quite cheap. I expect that wouldn't be worthwhile.

Reply:
I have finally figured out what my mistake was. I'm now confident in my benchmarks and what I'm seeing. I can finally start exploring when (or if) the partial heap sort is beneficial...

Reply:
Now trusting my numbers, I burned some cpu cycles to find the "break even" point. Turns out, the current implementation has a big constant factor:

```
BenchmarkDotNet v0.13.11, Windows 11 (10.0.22631.3007/23H2/2023Update/SunValley3)
AMD Ryzen 9 5900X, 1 CPU, 24 logical and 12 physical cores
.NET SDK 9.0.100-preview.2.24077.1
  [Host]     : .NET 9.0.0 (9.0.24.7604), X64 RyuJIT AVX2
  DefaultJob : .NET 9.0.0 (9.0.24.7604), X64 RyuJIT AVX
```

![image](https://github.com/dotnet/runtime/assets/16369295/bbb03b6a-012d-4a7e-9bcb-6a78f09d4ef7)

The benchmark does *not skip any elements* and always *takes 20 elements*, which should be a pretty favorable case for the "Top-N-Heapsort". 

Some further notes:
- As expected both Quick Sort and Top-N-Heapsort are highly sensitive to the data distribution. Above characteristics seem to representative tho
- Profiling shows `EnqueueDequeue` to be the hot path (as expected)

8K Elements for taking 20 amounts to about 0.2%. Do you think there is some potential in optimizing or may it not be worthwhile at all?


<details>
<summary>The code I used is more or less the same as in the PR, but for completness as I did it externally:</summary>

```
static IEnumerable<TElement> TakeOrdered<TElement, TKey>(IEnumerable<TElement> source, Func<TElement, TKey> keySelector, int minIdx, int maxIdx)
{
    int queueSize = maxIdx + 1;

    PriorityQueue<TElement, SortPriority<TKey>> priorityQueue = new(queueSize);

    int pos = 0;
    foreach (TElement element in source)
    {
        SortPriority<TKey> priority = new(keySelector(element), pos++);

        if (priorityQueue.Count == queueSize)
        {
            priorityQueue.EnqueueDequeue(element, priority);
        }
        else
        {
            priorityQueue.Enqueue(element, priority);
        }
    }
    TElement[] result = new TElement[priorityQueue.Count];

    for (int i = result.Length - 1; i >= 0; i--)
    {
        result[i] = priorityQueue.Dequeue();
    }

    for (int i = minIdx; i < result.Length; i++)
    {
        yield return result[i];
    }
}

internal readonly struct SortPriority<TKey> : IComparable<SortPriority<TKey>>
{
    private readonly TKey _key;
    private readonly int _sourcePos;

    public SortPriority(TKey key, int sourcePos) : this()
    {
        _key = key;
        _sourcePos = sourcePos;
    }

    public int CompareTo(SortPriority<TKey> other)
    {
        int result = -Comparer<TKey>.Default.Compare(_key, other._key);

        if (result == 0)
        {
            result = _sourcePos.CompareTo(other._sourcePos);
        }
        return result;
    }
}
```
</summary>